### ISSUE TITLE
garden running in bpm does not find/delete existing containers on startup

### ISSUE THREAD

**Issue by sunjayBhatia** [cloudfoundry/garden-runc-release#120]
## Description
When running garden in bpm, any orphaned containers are not cleaned up on restarts of the garden job. See reproduction steps below.

This affected us in our CI environment as these orphaned containers left state in the form of container network interfaces/network namespace files on the host VM. Subsequent container creations failed during network configuration b/c of this polluted state, causing Diego cells in CF to become unhealthy periodically.

## Links
- Diego investigation story: https://www.pivotaltracker.com/story/show/163212244
- Slack conversation: https://cloudfoundry.slack.com/archives/C0JV2NPJ4/p1547576166061500
- CF networking story: https://www.pivotaltracker.com/story/show/163266092

## Environment
* garden-runc-release version: `1.17.2`
* IaaS: bosh-lite + GCP
* Stemcell version: N/A
* Kernel version: N/A

## Steps to reproduce
- Deploy CF on bosh-lite with one Diego cell and BPM/rootless enabled
- Push an app
- Run `bosh stop diego-cell` (or `monit stop garden` from the cell)
- SSH onto the cell and see there is a `garden-init` process still running from the stranded application instance
- Run `bosh start diego-cell` (or `monit start garden` from the cell)
- SSH onto the cell and eventually see there are now two `garden-init` processes, the original stranded application (now pid 1) and another for the rescheduled app
- Look at the garden depot directory and see there is only one container in garden's record
- Look through the garden logs and see that there was no `clean-up-container` log line corresponding to the stranded container
- Look at output of `ifconfig` and see there are two container network interfaces

**Note:** stopping garden seems to only orphan 1 extra `garden-init` process if done repeatedly but you can see from the stranded container network interfaces that garden failed to fully delete many containers

## Cause
- Garden running in BPM does not keep its container depot state through restarts of the job so it loses track of containers that are running when the job exits

## Resolution
- Garden uses a volume mounted on the host to store container state so it can remember it between restarts?
- Other options pending

**Comment by BooleanCat** [comment]
We observe that when garden is stopped via `monit stop garden`, all of Garden's containers also stop. We appear to leak image dirs since we fail to create containers with the same name after restarting garden and the image plugin complains about this.

We also observe that the depot contains no trace of our container.

We can see from BPM code and logs that BPM attempts the following on gdn:

1. send sigterm
2. send sigquit
3. runc delete -f the bpm container
3.1 runc sends kill --all to the container

Outside of bpm in CF, garden is configured to destroy containers on start up. However, when garden dies, the container continue to function as normal, garden kills the containers next time it starts up.

Within BPM, because of the above, garden containers die when the garden bpm container dies.

Within a pid namespace, it is documented in man7 that when you kill the init process (pid 1), the kernel will send SIGKILL to all other processes in the namespace. Experimentally, we've observed that this applies to all nested child pid namespaces.

The conclusion is that, if we wish to continue to use BPM, then it is expected behaviour for containers to die when garden dies.

TODOs

1. Consider always enabling destroy_on_startup in rootless
2. Persist the depot in bpm mode
3. raise issue in BPM to document this "gotcha"
4. PR man7 to document the behaviour we obvserved (around pid namespace ancestors dying)
